### PR TITLE
Feature/make webapistartupfactory test services customizable

### PR DESCRIPTION
**Pull request by TomMalow**
This is a continuation of PR #23 

This PR adds the option to modify the test services in `WebApiStartupFactory` with a new partial method `ModifyServices()`. Thereby, providing the option to add mocked services.

### CODE REVIEW

File: src/Atc.Rest.ApiGenerator/Generators/ServerHostGenerator.cs [hunk 1]
```diff
@@ -889,6 +898,26 @@ private static MemberDeclarationSyntax CreateWebApplicationFactoryModifyConfigur
...
 
+        private static MemberDeclarationSyntax CreateWebApplicationFactoryModifyServices()
+        {
+            return SyntaxFactory.MethodDeclaration(
+                    SyntaxFactory.PredefinedType(
+                        SyntaxFactory.Token(SyntaxKind.VoidKeyword)),
+                    SyntaxFactory.Identifier("ModifyServices"))
+                .WithModifiers(
+                    SyntaxFactory.TokenList(
+                        SyntaxFactory.Token(SyntaxKind.PartialKeyword)))
```
Comment (davidkallesen) [inline src/Atc.Rest.ApiGenerator/Generators/ServerHostGenerator.cs:909]: Use SyntaxTokenFactory.PartialKeyword

File: src/Atc.Rest.ApiGenerator/Generators/ServerHostGenerator.cs [hunk 2]
```diff
@@ -889,6 +898,26 @@ private static MemberDeclarationSyntax CreateWebApplicationFactoryModifyConfigur
                     SyntaxFactory.Token(SyntaxKind.SemicolonToken));
         }
 
+        private static MemberDeclarationSyntax CreateWebApplicationFactoryModifyServices()
+        {
+            return SyntaxFactory.MethodDeclaration(
+                    SyntaxFactory.PredefinedType(
+                        SyntaxFactory.Token(SyntaxKind.VoidKeyword)),
```
Comment (davidkallesen) [inline src/Atc.Rest.ApiGenerator/Generators/ServerHostGenerator.cs:905]: Use SyntaxTokenFactory.VoidKeyword

File: src/Atc.Rest.ApiGenerator/Generators/ServerHostGenerator.cs [hunk 3]
```diff
@@ -889,6 +898,26 @@ private static MemberDeclarationSyntax CreateWebApplicationFactoryModifyConfigur
...
+                        SyntaxFactory.Token(SyntaxKind.PartialKeyword)))
+                .WithParameterList(
+                    SyntaxFactory.ParameterList(
+                        SyntaxFactory.SingletonSeparatedList(
+                            SyntaxFactory.Parameter(
+                                SyntaxFactory.Identifier("services"))
+                            .WithType(
+                                SyntaxFactory.IdentifierName("IServiceCollection")))))
+                .WithSemicolonToken(
+                    SyntaxFactory.Token(SyntaxKind.SemicolonToken));
```
Comment (davidkallesen) [inline src/Atc.Rest.ApiGenerator/Generators/ServerHostGenerator.cs:918]: Use SyntaxTokenFactory.Semicolon

File: src/Atc.Rest.ApiGenerator/Generators/ServerHostGenerator.cs [hunk 4]
```diff
@@ -889,6 +898,26 @@ private static MemberDeclarationSyntax CreateWebApplicationFactoryModifyConfigur
...
+                .WithModifiers(
+                    SyntaxFactory.TokenList(
+                        SyntaxFactory.Token(SyntaxKind.PartialKeyword)))
+                .WithParameterList(
+                    SyntaxFactory.ParameterList(
+                        SyntaxFactory.SingletonSeparatedList(
+                            SyntaxFactory.Parameter(
+                                SyntaxFactory.Identifier("services"))
+                            .WithType(
+                                SyntaxFactory.IdentifierName("IServiceCollection")))))
```
Comment (davidkallesen) [inline src/Atc.Rest.ApiGenerator/Generators/ServerHostGenerator.cs:916]: Group to previews line

File: src/Atc.Rest.ApiGenerator/Generators/ServerHostGenerator.cs [hunk 5]
```diff
@@ -889,6 +898,26 @@ private static MemberDeclarationSyntax CreateWebApplicationFactoryModifyConfigur
...
+                    SyntaxFactory.PredefinedType(
+                        SyntaxFactory.Token(SyntaxKind.VoidKeyword)),
+                    SyntaxFactory.Identifier("ModifyServices"))
+                .WithModifiers(
+                    SyntaxFactory.TokenList(
+                        SyntaxFactory.Token(SyntaxKind.PartialKeyword)))
+                .WithParameterList(
+                    SyntaxFactory.ParameterList(
+                        SyntaxFactory.SingletonSeparatedList(
+                            SyntaxFactory.Parameter(
```
Comment (davidkallesen) [inline src/Atc.Rest.ApiGenerator/Generators/ServerHostGenerator.cs:913]: Group to previews line

File: src/Atc.Rest.ApiGenerator/Generators/ServerHostGenerator.cs [hunk 6]
```diff
@@ -810,6 +811,14 @@ private static MemberDeclarationSyntax CreateWebApplicationFactoryConfigureWebHo
                                                 SyntaxFactory.Identifier("services")))
                                         .WithBlock(
                                             SyntaxFactory.Block(
+                                                SyntaxFactory.ExpressionStatement(
+                                                   SyntaxFactory.InvocationExpression(
+                                                       SyntaxFactory.IdentifierName("ModifyServices"))
+                                                   .WithArgumentList(
+                                                       SyntaxFactory.ArgumentList(
+                                                           SyntaxFactory.SingletonSeparatedList(
+                                                               SyntaxFactory.Argument(
```
Comment (davidkallesen) [inline src/Atc.Rest.ApiGenerator/Generators/ServerHostGenerator.cs:820]: Use SyntaxArgumentFactory.Create("services")

File: src/Atc.Rest.ApiGenerator/Generators/ServerHostGenerator.cs [hunk 7]
```diff
@@ -810,6 +811,14 @@ private static MemberDeclarationSyntax CreateWebApplicationFactoryConfigureWebHo
                                                 SyntaxFactory.Identifier("services")))
                                         .WithBlock(
                                             SyntaxFactory.Block(
+                                                SyntaxFactory.ExpressionStatement(
+                                                   SyntaxFactory.InvocationExpression(
+                                                       SyntaxFactory.IdentifierName("ModifyServices"))
+                                                   .WithArgumentList(
+                                                       SyntaxFactory.ArgumentList(
```
Comment (davidkallesen) [inline src/Atc.Rest.ApiGenerator/Generators/ServerHostGenerator.cs:818]: Use SyntaxArgumentListFactory.CreateWithOneItem("services")

**Comment by davidkallesen** [comment]
@TomMalow please look at review feedback

**Comment by TomMalow** [comment]
I will take a look at it. 

Note that the implementation is based on the existing solution added with PR #23. Should I update that implementation as well based on this feedback?

**Comment by TomMalow** [comment]
@davidkallesen The PR have been updated based on your feedback

File: src/Atc.Rest.ApiGenerator/Generators/ServerHostGenerator.cs [hunk 8]
```diff
@@ -638,7 +638,7 @@ private static MemberDeclarationSyntax CreateStartupPropertyPublicConfiguration(
                     SyntaxFactory.Identifier("Configuration"))
                 .WithModifiers(
                     SyntaxFactory.TokenList(
-                        SyntaxFactory.Token(SyntaxKind.PublicKeyword)))
+                        SyntaxTokenFactory.PublicKeyword()))
```
Comment (perkops) [inline src/Atc.Rest.ApiGenerator/Generators/ServerHostGenerator.cs:641]: I would move this line 1 up

**Pull request merged**